### DISCUSSION TITLE
Handling built in functions that aren't keywords

### DISCUSSION THREAD

**Discussion by katawful** [tree-sitter/tree-sitter#1426] [Q&A] [answered]
The language I am making a parser for has a lot of builtin functions that aren't really language keywords (such as `if`, `while`, etc...). They are static, but also cannot be overloaded by other objects so the following is invalid:

```
int x ; x is a valid name
int GetFPS ; GetFPS is already used by language
if (x == 1)
  set x = GetFPS ; x can be set to the function GetFPS
  set GetFPS = GetFPS ; since GetFPS is already use, you're setting a function to a function which breaks
endif
```
Now while this might not seem like a job not for tree-sitter, the entire language is built around these builtin functions that aren't keywords. Thus, in order to be able to build an accurate AST for the file for queries (mostly just highlighting), I would need a system to be able to detect these set off names when parsing the file. I have tried including them in the grammar file, but this is not only completely inefficient to generate due to the sheer number of these functions (taking upwards of 20GB of memory and 30 minutes on my system), but tree-sitter can't even distiguish these in the end anyways. I thought about using a scanner to do this, but I'm not sure how to do this either in a way that's efficient

**Comment by ahelwer** [comment]
What do you want tree-sitter's behavior to be when parsing your code snippet? You want it to output like a `built_in_function` node when it sees `GetFPS` and others? Or do you want the parsing to fail? It's possible to write an external scanner that basically functions as an [EXCLUDE rule](https://github.com/tree-sitter/tree-sitter/pull/246), but that is moving semantic checking into the syntactic level which you want to avoid if your only objective is highlighting.

I think you've basically tried to do this in the tree-sitter javascript DSL without success, so an external scanner is your next-best bet. You can add all your reserved keywords (normalized to lowercase) to a static `std::set<std::string>`, then set up your grammar so it first asks your external scanner whether something is a reserved keyword; then, in your external scanner, read ahead until you reach a space or some other non-alphanumeric character, normalize the read string to lowercase, then check for membership in your static keyword set. If it's present, emit a `BUILT_IN_FUNCTION` token.

Note that I don't think `std::set` is available in the tree-sitter webassembly build so your generated wasm won't work, but it will work in other contexts that just compile the C/C++ code. See [here](https://github.com/tree-sitter/tree-sitter/discussions/1325) for more info on this. If you need the webassembly build to work, maybe put your keywords in a `std::vector` in alphabetical order then do a binary search over it. You could also use something like a [prefix trie](https://en.wikipedia.org/wiki/Trie).

**Reply by katawful** [reply]
To be a bit more clear: I'm just looking, at the end of parsing, for these builtin functions in the AST to appear as an element in the tree. Using a choice DSL in the grammar file itself caused the problem I talked about

For your way to do it in the scanner, do you think it would be fast enough for parsing?

**Reply by ahelwer** [reply]
@katawful I think it probably would, a static set with only 3000-ish elements is very small and queries would be extremely fast. It's *possible* you could beat `std::set` by writing a prefix trie but I wouldn't bet on it. You'd probably have to do weird stuff like statically encoding offsets or whatever. One other option is to generate the matching code with a lexer generator. If you go this route please do report results, I am interested to see how this approach works.

**Reply by katawful** [reply]
Alright, thanks. I'll have to poke around with it

**Reply by ahelwer** [reply]
I actually found [this pdf](http://www.lafstern.org/matt/col1.pdf) which claims that binary search over a sorted vector is superior to `std::set` in terms of both time & memory when you don't care about insertion performance (which you definitely don't, because it's a static list). So that's another option. But this is premature optimization, I doubt the difference will be noticeable.

**Comment by maxbrunsfeld** [comment, marked as answer]
You should handle this outside of your Tree-sitter parser. People often make the mistake of thinking that every error needs to be detected at parse time. It's better to simply parse these nodes as an `identifier`. Then, in your own code do whatever *semantic* correctness checks you want, by walking the tree and checking things.

**Reply by katawful** [reply]
My issue is that doesn't lead to as many semantic problems is that it does syntax ones. It's not just that these functions remove possible names from the user, it's that they also directly distinguish syntax that otherwise has identical regex. Not being able to detect these would make tree-sitter as a parser functionally useless over regex based alternatives. These functions are not called with their own unique syntax, there is an overlap with multiple other objects which leads to ambiguity otherwise. Being able to detect these would immediately remove any ambiguity
